python/pyarrow/parquet/core.py

```diff
@@ -3423,16 +3427,22 @@ def write_metadata(schema, where, metadata_collector=None, **kwargs):
     ...     table.schema, 'dataset_metadata/_metadata',
     ...     metadata_collector=metadata_collector)
     """
-    writer = ParquetWriter(where, schema, **kwargs)
+    filesystem, where = _resolve_filesystem_and_path(where, filesystem)
+
+    writer = ParquetWriter(where, schema, filesystem, **kwargs)
     writer.close()
 
     if metadata_collector is not None:
         # ParquetWriter doesn't expose the metadata until it's written. Write
         # it and read it again.
-        metadata = read_metadata(where)
+        metadata = read_metadata(where, filesystem=filesystem)
         for m in metadata_collector:
             metadata.append_row_groups(m)
-        metadata.write_metadata_file(where)
+        file_ctx = nullcontext()
+        if filesystem is not None:
+            file_ctx = where = filesystem.open_output_stream(where)
+        with file_ctx:
+            metadata.write_metadata_file(where)
```

Comment:
```suggestion
        if filesystem is not None:
            with filesystem.open_output_stream(where) as dest:
                metadata.write_metadata_file(dest)
        else:
            metadata.write_metadata_file(where)
```

Just a nit, but to see the diff, this is just as many lines of code, and even less characters, while trying to be less "smart" (and IMO easier to read)

Reply:
Actually, one more thing that I realize now: with using `_resolve_filesystem_and_path`, the only case when `filesystem` will still be `None`, is when `where` is a file-like object. But in that case, I am not sure the code to handle `metadata_collector` would work anyway? (like writing, reading and writing again to the same file like object, will that work?) It might be good to test that, so we either have coverage for it (if it works), or either we can simplify the above code (since we can assume `filesystem` to be always None)

Reply:
> Just a nit, but to see the diff, this is just as many lines of code, and even less characters, while trying to be less "smart" (and IMO easier to read)

Ya, I sort of agree, but saw read_meta used that pattern and thought maybe that was preferred.

Reply:
Didn't work out of the box with file-like, but some small modifications for setting cursor back after reading before re-writing seems to work okay. https://github.com/apache/arrow/pull/14574/commits/54dd6415189ae8fe475e8965f949130b904065d1